### PR TITLE
feat: verify hot node

### PR DESCRIPTION
**Pull request by swelf19**
Implemented full signature verification on a hot node.
added verify_batch command to prysm_compability_checker
```bash
./dc4bc_prysm_compatibility_checker verify_batch /tmp/swelf/dkg_signatures_dump_a7a26.json 'mWkXWHsaqcGbmCqcGEn9vnLkVS+df54mzF3nxd6ObDF6Mvr2Hs1rThjYPkSGllM8' /tmp/messages/
All batch signatures are correct
```

in case of failed verisifation we get an error
```
./dc4bc_prysm_compatibility_checker verify_batch /tmp/swelf/dkg_signatures_dump_a7a26.json 'mWkXWHsaqcGbmCqcGEn9vnLkVS+df54mzF3nxd6ObDF6Mvr2Hs1rThjYPkSGllM8' /tmp/messages/
2022/01/24 16:41:57 failed to verify prysm signature for file - file1
```

### CODE REVIEW

**Review by pr0n00gler — COMMENTED** [review]
Pls run ```go mod tidy``` and ```goimports -w...```

**Comment by oopcode** [comment]
due to a lack of a clear description of the task (this is my fault) the contents of this PR do not solve our problems.

We have this guy here: https://github.com/lidofinance/dc4bc/tree/master/cmd/prysm_compatibility_checker, and we have the export_signatures command that exports the signatures to a file. After they have been exported, we need a way to make sure that the signatures are correct. We need to make the compatibility checker work with that export format.

File: cmd/prysm_compatibility_checker/main.go [hunk 1]
```diff
@@ -80,6 +83,37 @@ func verify() *cobra.Command {
 	}
 }
 
+func verifyBatch() *cobra.Command {
+	return &cobra.Command{
+		Use:   "verify_batch [exported_signatures] [pubkey] [dir]",
```
Comment (sotnikov-s) [inline cmd/prysm_compatibility_checker/main.go:88]: at first glance, I didn't get what `exported_signatures` is about. there is also no mention of it in the description of the command. I suggest to make it more obvious how to use the command, e.g. expand Short of the command with `./dc4bc_cli export_signatures` mention

**Comment by oopcode** [comment]
also please add a readme update with a step by step description of the verification process

File: client/flow_test.go [hunk 2]
```diff
@@ -827,6 +880,9 @@ func testReinitDKGFlow(t *testing.T, convertDKGTo10_1_4 bool) {
 	numNodes := 4
 	threshold := 2
 	startingPort := 8095
+	if convertDKGTo10_1_4 {
+		startingPort = 8100
```
[the remark this answers is not in the document]
Comment (swelf19) [reply client/flow_test.go:1061]: just follow the previous tests convention, may be to run tests in parallel. nonoverlapping port range for each test

File: client/services/node/node_service.go [hunk 3]
```diff
@@ -867,3 +878,27 @@ func recoverFullSign(signingFSM *state_machines.FSMInstance, msg []byte, sigShar
 	}
 	return tbls.Recover(suite.(pairing.Suite), blsKeyring.PubPoly, msg, sigShares, t, n)
 }
+
+// VerifySign verifies a signature of a message
+func VerifySign(signingFSM *state_machines.FSMInstance, msg []byte, fullSignature []byte) error {
```
[the remark this answers is not in the document]
Comment (swelf19) [reply client/services/node/node_service.go]: dead code from previous iteration

File: client/services/signature/signature.go [hunk 4]
```diff
@@ -1,15 +1,26 @@
...
 )
 
 type SignatureService interface {
 	GetSignatures(dto *dto.DkgIdDTO) (map[string][]types.ReconstructedSignature, error)
 	GetSignatureByID(dto *dto.SignatureByIdDTO) ([]types.ReconstructedSignature, error)
-	SaveSignatures(signature []types.ReconstructedSignature) error
+	GetSignaturesByBatchID(dto *dto.SignaturesByBatchIdDTO) (map[string][]types.ReconstructedSignature, error)
+	GetBatches(dto *dto.DkgIdDTO) (map[string][]string, error)
+	SaveSignatures(batchID string, signature []types.ReconstructedSignature) error
+	VerifySign(signingFSM *state_machines.FSMInstance, dto *dto.SignatureByIdDTO) error
```
[the remark this answers is not in the document]
Comment (swelf19) [reply client/services/signature/signature.go]: dead code from previous iteration

File: client/repositories/signature/signature.go [hunk 5]
```diff
@@ -96,9 +132,49 @@ func (r *BaseSignatureRepo) SaveSignatures(signaturesToSave []types.Reconstructe
 
 	key := state.MakeCompositeKeyString(SignaturesKeyPrefix, signaturesToSave[0].DKGRoundID)
 
+	// WARN: ideally we have to make both save operations in single transaction
```
[the remark this answers is not in the document]
Comment (swelf19) [reply client/repositories/signature/signature.go]: at first sight two models for batch signatures looks ok. But you are right one nested map is easier to implement and maintain.

**Review by oopcode — APPROVED** [review]

**Review by sotnikov-s — APPROVED** [review]

File: client/repositories/signature/signature.go [hunk 6]
```diff
@@ -13,10 +13,37 @@ const (
 	SignaturesKeyPrefix = "signatures"
 )
 
+type SignaturesStorage map[string]map[string][]types.ReconstructedSignature
+
+func (s *SignaturesStorage) AddReconstructedSignature(reconstructedSignature types.ReconstructedSignature) {
```
[the remark this answers is not in the document]
Comment (swelf19) [reply client/repositories/signature/signature.go:18]: i kept this in mind, but i thought its a good idea to show it is a readwrite method.

**Review by oopcode — APPROVED** [review]

**Pull request merged**